File: 0070-climbing-stairs/0070-climbing-stairs.js
/**
 * @param {number} n
 * @return {number}
 */

let climb = (numStairs, memo = {}) => {
    if(numStairs in memo)
        return memo[numStairs]
    
    if(numStairs === 0)
        return 1
    if(numStairs < 0)
        return 0;
    
    let totalStepsToTake = 0
    if(numStairs - 1 in memo)
        totalStepsToTake += memo[numStairs - 1]
    else 
        totalStepsToTake += climb(numStairs - 1, memo)

    
    if(numStairs - 2 in memo)
        totalStepsToTake += memo[numStairs - 2]
    else 
        totalStepsToTake += climb(numStairs - 2, memo)

    memo[numStairs] = totalStepsToTake
    return memo[numStairs]
}

var climbStairs = function(n) {
//     from each step, we can go either 1 step or 2 steps
   return climb(n)

};

File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {
    int[] dp;
    
    public int getNumWays(int n) {
        if(n == 0)
            return 1;
        
        if(n < 0)
            return 0;
        
        if(dp[n] != -1)
            return dp[n];
        
        dp[n] = getNumWays(n-1) + getNumWays(n-2);
        return dp[n];
    }
    
    public int climbStairs(int n) {
        dp = new int[n+1];
        for(int i=0; i<=n; i++)
            dp[i] = -1;
        
        return getNumWays(n);
    }
}